app/src/org/commcare/dalvik/activities/PinAuthenticationActivity.java

```diff
@@ -0,0 +1,173 @@
+package org.commcare.dalvik.activities;
+
+import android.content.Intent;
+import android.os.Bundle;
+import android.util.Log;
+import android.view.View;
+import android.widget.Button;
+import android.widget.EditText;
+import android.widget.TextView;
+import android.widget.Toast;
+
+import org.commcare.android.database.app.models.UserKeyRecord;
+import org.commcare.android.framework.ManagedUi;
+import org.commcare.android.framework.SessionAwareCommCareActivity;
+import org.commcare.android.framework.UiElement;
+import org.commcare.dalvik.R;
+import org.commcare.dalvik.application.CommCareApplication;
+import org.javarosa.core.services.locale.Localization;
+
+/**
+ * When a user selects Set/Reset PIN from the home activity options menu, they must first
```

Comment:
Responded in the other place. We have access to the user's password in the local context for the length of their session so we can perform web requests with it, so we should just use that cached value.